### PR TITLE
feat(components): Drawer component

### PR DESCRIPTION
**Pull request by bdfranck**
This PR adds a drawer component based on the [previously published proposal](https://github.com/GetJobber/atlantis/tree/master/docs/proposals/Drawer).

drawer-toggle

### Added

- Added a drawer component.

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

**Comment by bdfranck** [comment]
My main uncertainty is how the drawer interacts with the layout of the parent element. Here are some options:

We could assume that the drawer is in a flex layout such as the following:

```
<div>
  <div>Main content</div>
  <Drawer title="A Drawer">
    Drawer content
  </Drawer>
</div>
```

Or we could have the drawer provide a wrapper for the main content such as the following:

```
<DrawerWrap>
  <div>Main content</div>
  <Drawer title="A Drawer">
    Drawer content
  </Drawer>
</div>
```

Or something else?

**Comment by bdfranck** [comment]
Here's what's incomplete:

- This assumes we have access to a `<grid>` component like my proposal. I've added some inline CSS to the `mdx` file to show the effect.

Here's where I'm stuck:

- I'm not sure how to disable the page scrolling when the drawer fills the screen for mobile
- I'm not sure how to specify an external activator button to open the drawer

**Comment by eddysims** [comment]
@seanhealy what are your thoughts on doing a drawer similar to the way we did `Toast`? This would mean that we don't need the `Page` or `Grid` component, as the drawer would be rendered at the end of the dom.

This would allow the drawer to be worked on right now as it would limit dependencies. However I'm not sure if this is a route that we want to go to often.

@bdfranck for reference. When we use the `Toast` component, we create a new small React App on the page that renders the Toast. This allows for it to be used multiple times on a page, or share a toast between React Apps without needing a `Provider`

**Comment by seanhealy** [comment]
@eddysims I don't mind that idea. We would still need an anchor point for the drawer though as it does push the main content over when it opens.

**Comment by bdfranck** [comment]
@eddysims That's interesting but it may be overkill for this. I chatted with @JeffMarvin last week and he identified two different types of drawers: an app-level one and a page-level one. The ML team only needs a page-level one at this point. Perhaps I could name it accordingly?

image

**Comment by chris-at-jobber** [comment]
@bdfranck we've fixed an issue causing specs to fail - if you update this branch from the latest master specs will pass ✅

**Comment by bdfranck** [comment]
@chris-at-jobber Thanks for the advice! Is there a process to flag this PR as ready for an external review again?

**Comment by chris-at-jobber** [comment]
I think if we flip the `review requested` label it will ping McWeb

**Comment by chris-at-jobber** [comment]
I was wrong! Go to the "Reviewers" section, remove McWeb as a reviewer, and add McWeb back! (In the meantime, we are aware of this pending review 😄 )

image

**Comment by bdfranck** [comment]
@chris-at-jobber This is ready to review again! 🥳

File: packages/components/src/Drawer/Drawer.mdx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+Drawers are temporary sidebars that allow users to view supplemantary content.
+Users can interact with the main content while the drawer is visible.
+
+```ts
+import { Page } from "@jobber/components/Page";
+```
+
+<Playground>
+  {() => {
+    const [modalOpen, setModalOpen] = useState(true);
```
Comment (seanhealy) [inline packages/components/src/Drawer/Drawer.mdx:30]: `drawerOpen` may be better here.

File: packages/components/src/Drawer/Drawer.test.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+import renderer from "react-test-renderer";
+import { cleanup } from "@testing-library/react";
+import { Drawer } from ".";
+
+afterEach(cleanup);
+
+it("renders a Drawer", () => {
```
Comment (seanhealy) [inline packages/components/src/Drawer/Drawer.test.tsx:8]: Would probably be good to test that it doesn't render a drawer if `open` is false.

Comment (seanhealy) [reply packages/components/src/Drawer/Drawer.test.tsx:8]: Also, if that `onRequestClose` is called when the `✖️` is clicked.

File: packages/components/src/Drawer/Drawer.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import { Icon } from "../Icon";
+
+interface DrawerProps {
+  readonly children: ReactNode | ReactNode[];
+  readonly title: string;
+  /**
+   * @default true
+   */
+  readonly open?: boolean;
+  onRequestClose?(): void;
```
Comment (seanhealy) [inline packages/components/src/Drawer/Drawer.tsx:14]: I think we _may_ want to make this required as if they don't hook it up the `✖️` will do nothing when clicked.
```suggestion
  onRequestClose(): void;
```

File: packages/components/src/Drawer/Drawer.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    </>
+  );
+}
+
+interface HeaderProps {
+  title: string;
+  onRequestClose?(): void;
+}
+
+function Header({ title, onRequestClose }: HeaderProps) {
```
Comment (seanhealy) [inline packages/components/src/Drawer/Drawer.tsx:49]: Like that you split this out!

**Review by seanhealy — CHANGES_REQUESTED** [review]
@bdfranck Looking good! Just had a couple little thoughts for you. Lemme know if you have any questions.

**Comment by bdfranck** [comment]
### Cross-browser testing

#### Chrome (Latest)

- [ ] Displays properly on small breakpoint
- [ ] Displays properly on medium / large breakpoints
- [ ] Drawer toggles open and closed

#### Firefox (Latest)

- [ ] Displays properly on small breakpoint
- [ ] Displays properly on medium / large breakpoints
- [ ] Drawer toggles open and closed


#### Safari (Latest)

- [ ] Displays properly on small breakpoint
- [ ] Displays properly on medium / large breakpoints
- [ ] Drawer toggles open and closed


#### Edge (Latest)

- [ ] Displays properly on small breakpoint
- [ ] Displays properly on medium / large breakpoints
- [ ] Drawer toggles open and closed


#### Chrome on Android (Latest)

- [ ] Displays properly on small breakpoint
- [ ] Displays properly on medium / large breakpoints
- [ ] Drawer toggles open and closed


#### Safari on iOS (Latest)

- [ ] Displays properly on small breakpoint
- [ ] Displays properly on medium / large breakpoints
- [ ] Drawer toggles open and closed

File: packages/components/src/Drawer/Drawer.mdx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+Drawers are temporary sidebars that allow users to view supplemantary content.
+Users can interact with the main content while the drawer is visible.
+
+```ts
+import { Page } from "@jobber/components/Page";
+```
+
+<Playground>
+  {() => {
+    const [drawerOpen, setModalOpen] = useState(true);
```
Comment (seanhealy) [inline packages/components/src/Drawer/Drawer.mdx:30]:
```suggestion
    const [drawerOpen, setDrawerOpen] = useState(true);
```

**Review by seanhealy — CHANGES_REQUESTED** [review]
One last little rename and I think I'm happy!

**Review by seanhealy — APPROVED** [review]

**Pull request merged**